# EBAC - Regressão II - regressão múltipla

## Tarefa I

#### Previsão de renda II

Vamos continuar trabalhando com a base 'previsao_de_renda.csv', que é a base do seu próximo projeto. Vamos usar os recursos que vimos até aqui nesta base.

|variavel|descrição|
|-|-|
|data_ref                | Data de referência de coleta das variáveis |
|index                   | Código de identificação do cliente|
|sexo                    | Sexo do cliente|
|posse_de_veiculo        | Indica se o cliente possui veículo|
|posse_de_imovel         | Indica se o cliente possui imóvel|
|qtd_filhos              | Quantidade de filhos do cliente|
|tipo_renda              | Tipo de renda do cliente|
|educacao                | Grau de instrução do cliente|
|estado_civil            | Estado civil do cliente|
|tipo_residencia         | Tipo de residência do cliente (própria, alugada etc)|
|idade                   | Idade do cliente|
|tempo_emprego           | Tempo no emprego atual|
|qt_pessoas_residencia   | Quantidade de pessoas que moram na residência|
|renda                   | Renda em reais|

In [1]:
# import math

import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

# from sklearn.model_selection import train_test_split
# from sklearn import metrics
# from sklearn.ensemble import RandomForestClassifier

# from scipy.stats import ks_2samp
import statsmodels.formula.api as smf
import statsmodels.api as sm
import patsy

%matplotlib inline

In [2]:
df = pd.read_csv('previsao_de_renda.csv')

In [3]:
df = df[['sexo', 'posse_de_veiculo', 'posse_de_imovel', 'qtd_filhos', 'tipo_renda', 'educacao', 'estado_civil',
        'tipo_residencia', 'idade', 'tempo_emprego', 'qt_pessoas_residencia', 'renda']]

In [10]:
df.dropna(inplace=True)

df = pd.get_dummies(df, columns=['sexo','tipo_renda', 'educacao', 'estado_civil','tipo_residencia'], drop_first=True)

1. Separe a base em treinamento e teste (25% para teste, 75% para treinamento).
2. Rode uma regularização *ridge* com alpha = [0, 0.001, 0.005, 0.01, 0.05, 0.1] e avalie o $R^2$ na base de testes. Qual o melhor modelo?
3. Faça o mesmo que no passo 2, com uma regressão *LASSO*. Qual método chega a um melhor resultado?
4. Rode um modelo *stepwise*. Avalie o $R^2$ na vase de testes. Qual o melhor resultado?
5. Compare os parâmetros e avalie eventuais diferenças. Qual modelo você acha o melhor de todos?
6. Partindo dos modelos que você ajustou, tente melhorar o $R^2$ na base de testes. Use a criatividade, veja se consegue inserir alguma transformação ou combinação de variáveis.
7. Ajuste uma árvore de regressão e veja se consegue um $R^2$ melhor com ela.

In [11]:
X = df.renda
y = df.drop('renda', axis=1)

In [12]:
from sklearn.model_selection import train_test_split

# dividindo a base em treinamento e teste
X_train, X_test, y_train, y_test = train_test_split(df.drop('renda', axis=1), df['renda'], test_size=0.25, random_state=42)


In [13]:
from sklearn.linear_model import Ridge
from sklearn.metrics import r2_score

alphas = [0, 0.001, 0.005, 0.01, 0.05, 0.1]

for alpha in alphas:
    model = Ridge(alpha=alpha)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    print(f"Alpha = {alpha:.3f}, R2 = {r2:.3f}")

Alpha = 0.000, R2 = 0.298
Alpha = 0.001, R2 = 0.298
Alpha = 0.005, R2 = 0.298
Alpha = 0.010, R2 = 0.298
Alpha = 0.050, R2 = 0.298
Alpha = 0.100, R2 = 0.298


In [15]:
from sklearn.linear_model import Lasso


for alpha in alphas:
    model = Lasso(alpha=alpha)
    model.fit(X_train, y_train)
    y_pred = model.predict(X_test)
    r2 = r2_score(y_test, y_pred)
    print(f"Alpha = {alpha:.3f}, R2 = {r2:.3f}")

C:\Users\joao3\AppData\Local\Temp/ipykernel_14724/2692951802.py:6: UserWarning: With alpha=0, this algorithm does not converge well. You are advised to use the LinearRegression estimator
  model.fit(X_train, y_train)
D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: UserWarning: Coordinate descent with no regularization may lead to unexpected results and is discouraged.
  model = cd_fast.enet_coordinate_descent(
D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 303648728790.6538, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(


Alpha = 0.000, R2 = 0.298
Alpha = 0.001, R2 = 0.298


D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 287595087998.9199, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(
D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 234333773242.55127, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(


Alpha = 0.005, R2 = 0.298
Alpha = 0.010, R2 = 0.298


D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 185806748508.84552, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(
D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 50014958477.56958, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(


Alpha = 0.050, R2 = 0.298
Alpha = 0.100, R2 = 0.298


D:\Arquivos de Programas\anaconda3\lib\site-packages\sklearn\linear_model\_coordinate_descent.py:530: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Duality gap: 17796395792.870056, tolerance: 80599627.18769649
  model = cd_fast.enet_coordinate_descent(


In [18]:
!pip install mlxtend

  Using cached joblib-1.2.0-py3-none-any.whl (297 kB)
  Attempting uninstall: joblib
    Found existing installation: joblib 1.1.0
    Uninstalling joblib-1.1.0:
      Successfully uninstalled joblib-1.1.0
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.24.2
    Uninstalling scikit-learn-0.24.2:
      Successfully uninstalled scikit-learn-0.24.2


In [24]:
from mlxtend.feature_selection import SequentialFeatureSelector
from sklearn.linear_model import LinearRegression


lr = LinearRegression()


sfs_stepwise = SequentialFeatureSelector(lr, forward=True, k_features=5)


X_selected = sfs_stepwise.fit_transform(X_train, y_train)


print("Índices das características selecionadas:")
print(sfs_stepwise.k_feature_idx_)


print("\nCaracterísticas selecionadas:")
print(sfs_stepwise.k_feature_names_)


Índices das características selecionadas:
(3, 4, 6, 8, 13)

Características selecionadas:
('idade', 'tempo_emprego', 'sexo_M', 'tipo_renda_Empresário', 'educacao_Superior completo')


TypeError: '(slice(None, None, None), (3, 4, 6, 8, 13))' is an invalid key

In [36]:
modelo = 'np.log(renda) ~ sexo_M + posse_de_veiculo + posse_de_imovel + tipo_renda_Empresário + tempo_emprego'
md = smf.ols(modelo, data = df)
reg = md.fit_regularized(method = 'elastic_net' 
                         , refit = True
                         , L1_wt = 1
                         , alpha = 0.01)

reg.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          np.log(renda)   R-squared:                       0.344
Model:                            OLS   Adj. R-squared:                  0.344
Method:                 Least Squares   F-statistic:                     1305.
Date:                Thu, 27 Apr 2023   Prob (F-statistic):               0.00
Time:                        20:25:39   Log-Likelihood:                -13694.
No. Observations:               12427   AIC:                         2.740e+04
Df Residuals:                   12422   BIC:                         2.744e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
============================================================================================
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept                    7.4177      0.015    482.233      0.000       7.388       7.448
posse_de_veiculo[T.True]     0.0530      0.014      3.771      0.000       0.025       0.081
posse_de_imovel[T.True]      0.0956      0.014      6.954      0.000       0.069       0.123
sexo_M                       0.7629      0.015     52.474      0.000       0.734       0.791
tipo_renda_Empresário             0          0        nan        nan           0           0
tempo_emprego                0.0629      0.001     64.026      0.000       0.061       0.065
==============================================================================
Omnibus:                        0.946   Durbin-Watson:                   2.024
Prob(Omnibus):                  0.623   Jarque-Bera (JB):                0.919
Skew:                           0.019   Prob(JB):                        0.631
Kurtosis:                       3.019   Cond. No.                         31.5
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [52]:
r_squared = reg.rsquared
print('O R² do modelo é:', r_squared)


O R² do modelo é: 0.3443041249285532


Foi separado a base em 25% para teste e 75% para treinamento. Rodmaos reg com ridge e o R² se manteve estático. A mesma coisa com LASSO e Stepwise. Foi ajustado o modelo com as melhores váriaveis, e o R² da árvore de regressão melhorou, ficando 34%.